diff -Naur uboot/examples/Makefile uboot.fc6/examples/Makefile
--- uboot/examples/Makefile	2005-12-16 11:39:27.000000000 -0500
+++ uboot.fc6/examples/Makefile	2006-11-07 15:17:29.227124931 -0500
@@ -150,6 +150,8 @@
 
 all:	$(obj).depend $(OBJS) $(LIB) $(SREC) $(BIN) $(ELF)
 
+$(SREC) : $(BIN)
+
 #########################################################################
 $(LIB):	$(obj).depend $(LIBOBJS)
 		$(AR) $(ARFLAGS) $@ $(LIBOBJS)
