--- uboot/arch/arm/cpu/arm1176/ast2500/platform.S.old	2016-12-29 10:13:52.000000000 +0800
+++ uboot/arch/arm/cpu/arm1176/ast2500/platform.S	2016-12-29 10:15:38.196052627 +0800
@@ -427,13 +427,13 @@
 /******************************************************************************
  Disable WDT2 for 2nd boot function
  ******************************************************************************/
-
+/* BMC 2nd boot is upon hardware strap bit17
 #ifndef CONFIG_FIRMWARE_2ND_BOOT
     ldr   r0, =0x1e78502c
     mov   r1, #0
     str   r1, [r0]
 #endif
-
+*/
 /******************************************************************************
  Disable WDT3 for SPI Address mode (3 or 4 bytes) detection function
  ******************************************************************************/
