--- u-boot-1.1.6/Makefile	2006-11-02 22:15:01.000000000 +0800
+++ u-boot-1.1.6-ami/Makefile	2009-08-12 11:10:30.000000000 +0800
@@ -1949,6 +1949,9 @@
 	@echo "#define CONFIG_GTH2 1" >>$(obj)include/config.h
 	@$(MKCONFIG) -a gth2 mips mips gth2
 
+AST2300EVB_config : unconfig
+	@./mkconfig $(@:_config=) arm arm926ejs AST2300EVB NULL AST2300
+
 #########################################################################
 ## S3C44B0 Systems
 #########################################################################
