diff -Naur u-boot-1.1.4/cpu/arm926ejs/config.mk u-boot-1.1.4-CPU/cpu/arm926ejs/config.mk
--- u-boot-1.1.4/cpu/arm926ejs/config.mk	2005-12-16 11:39:27.000000000 -0500
+++ u-boot-1.1.4-CPU/cpu/arm926ejs/config.mk	2006-05-18 14:06:32.000000000 -0400
 # =========================================================================
@@ -30,5 +30,9 @@
 # Supply options according to compiler version
 #
 # =========================================================================
-PLATFORM_CPPFLAGS +=$(call cc-option,-mapcs-32,-mabi=apcs-gnu)
+# As of now my compiler does not accept abi=apcs-gnu.
+#
+#PLATFORM_CPPFLAGS +=$(call cc-option,-mapcs-32,-mabi=apcs-gnu)
+PLATFORM_CPPFLAGS +=$(call cc-option,-mapcs-32)
+PLATFORM_CPPFLAGS +=-mno-thumb-interwork
 PLATFORM_RELFLAGS +=$(call cc-option,-mshort-load-bytes,$(call cc-option,-malignment-traps,))
