--- u-boot-1.1.6/include/ast2300/hwmap.h	1970-01-01 08:00:00.000000000 +0800
+++ u-boot-1.1.6-ami/include/ast2300/hwmap.h	2010-01-18 10:37:10.000000000 +0800
@@ -0,0 +1,76 @@
+/*
+ *  This file contains the AST2300 SOC Register Base, IRQ and DMA  mappings
+ *
+ *  Copyright (C) 2005 American Megatrends Inc
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+#ifndef     _HWREG_AST2300_H_
+#define     _HWMAP_AST2300_H_
+
+
+/********************************************************************
+   Internal Register Mapping Actual = 0x16000000 to 0x1E78CFFF
+   After Align the base and size    = 0x16000000 to 0x1EFFFFFF
+*********************************************************************/
+
+#define AST_REGISTER_BASE				0x16000000
+#define AST_REGISTER_SIZE				0x09000000
+
+/*--------------------------- Individual Devices ------------------------- */
+#define AST_SMC_BASE					0x16000000
+#define AST_AHBC_BASE                   0x1E600000
+#define AST_MAC1_BASE					0x1E660000
+#define AST_MAC2_BASE					0x1E680000
+#define AST_IC_BASE						0x1E6C0000
+#define AST_SDRAMC_BASE                 0x1E6E0000
+#define AST_SCU_BASE					0x1E6E2000
+#define AST_TIMER_BASE					0x1E782000
+#define AST_UART1_BASE                  0x1E783000
+#define AST_UART2_BASE                  0x1E784000
+#define AST_WDT_BASE					0x1E785000
+
+
+/*--------------- Virtual address of the IO Registers Region  ------------*/
+#define AST_REGISTER_VA_BASE			IO_ADDRESS(AST_REGISTER_BASE)
+
+#define AST_SMC_VA_BASE                 IO_ADDRESS(AST_SMC_BASE)
+#define AST_AHBC_VA_BASE                IO_ADDRESS(AST_AHBC_BASE)
+#define AST_MAC1_VA_BASE                IO_ADDRESS(AST_MAC1_BASE)
+#define AST_IC_VA_BASE					IO_ADDRESS(AST_IC_BASE)
+#define AST_SDRAMC_VA_BASE              IO_ADDRESS(AST_SDRAMC_BASE)
+#define AST_SCU_VA_BASE					IO_ADDRESS(AST_SCU_BASE)
+#define AST_TIMER_VA_BASE				IO_ADDRESS(AST_TIMER_BASE)
+#define AST_UART1_VA_BASE               IO_ADDRESS(AST_UART1_BASE)
+#define AST_UART2_VA_BASE               IO_ADDRESS(AST_UART2_BASE)
+#define AST_WDT_VA_BASE					IO_ADDRESS(AST_WDT_BASE)
+
+
+/*****************************************************************
+						    Flash
+*****************************************************************/
+#define CPE_FLASH_BASE                  0x10000000
+#define CPE_FLASH_SZ                    0x00800000		/* Max 16 MB */
+#define CPE_FLASH_VA_BASE               MEM_ADDRESS(CPE_FLASH_BASE)
+
+
+/*****************************************************************
+					  IRQ Assignment
+*****************************************************************/
+#define NR_IRQS							32
+#define MAXIRQNUM                       31
+
+#endif
