diff -Naur linux/include/uapi/linux/ethtool.h linux.new/include/uapi/linux/ethtool.h
--- linux/include/uapi/linux/ethtool.h	2017-09-25 14:52:28.234704111 +0800
+++ linux.new/include/uapi/linux/ethtool.h	2017-09-25 15:37:41.546623784 +0800
@@ -933,6 +933,7 @@
 #define SUPPORTED_40000baseCR4_Full	(1 << 24)
 #define SUPPORTED_40000baseSR4_Full	(1 << 25)
 #define SUPPORTED_40000baseLR4_Full	(1 << 26)
+#define SUPPORTED_25000baseT_Full     (1 << 27)
 
 /* Indicates what features are advertised by the interface. */
 #define ADVERTISED_10baseT_Half		(1 << 0)
@@ -975,6 +976,7 @@
 #define SPEED_1000		1000
 #define SPEED_2500		2500
 #define SPEED_10000		10000
+#define SPEED_25000		25000
 #define SPEED_UNKNOWN		-1
 
 /* Duplex, half or full. */
