--- busybox.old/networking/udhcp/dhcpc.h	2008-11-10 01:27:58.000000000 +0800
+++ busybox.new/networking/udhcp/dhcpc.h	2010-05-12 09:48:36.000000000 +0800
@@ -23,6 +23,7 @@
 	uint8_t *vendorclass;           /* Optional vendor class-id to use */
 	uint8_t *hostname;              /* Optional hostname to use */
 	uint8_t *fqdn;                  /* Optional fully qualified domain name to use */
+    uint8_t interval;               /* Optional interval value for retry */
 };
 
 /* server_config sits in 1st half of bb_common_bufsiz1 */
