--- busybox-old/networking/ifupdown.c	2010-05-23 00:07:00.676974967 +0800
+++ busybox-new/networking/ifupdown.c	2010-05-25 06:22:11.359127024 +0800
@@ -372,8 +372,12 @@
 static int dhcp_up6(struct interface_defn_t *ifd, execfn *exec)
 {
 	int result = 0;
-	result = execute("dhcp6c -p /var/run/dhcp6c.%iface%.pid %iface% -f &", ifd, exec);
-	return (result);
+
+    if (!(result = execute("dhcp6.script %iface%", ifd, exec))){
+    	return execute("dhcp6c -p /var/run/dhcp6c.%iface%.pid %iface% ", ifd, exec);
+    }else{
+        return (result);
+    }
 }
 
 static int dhcp_down6(struct interface_defn_t *ifd, execfn *exec)
@@ -502,6 +506,10 @@
 		"pump -i %iface%[[ -h %hostname%]][[ -l %leasehours%]]",
 		"pump -i %iface% -k",
 	},
+    { "dhcp4.script",    //AMI dhcp script
+        "dhcp4.script  %iface% [[-h %hostname%]] [[-l %clientid%]]",
+        "kill `cat /var/run/udhcpc.%iface%.pid` 2>/dev/null",
+    },    
 	{ "udhcpc",
 		"udhcpc -R -n -p /var/run/udhcpc.%iface%.pid -i %iface%[[ -H %hostname%]][[ -c %clientid%]]"
 				"[[ -s %script%]][[ %udhcpc_opts%]]",
@@ -533,6 +541,8 @@
 #elif ENABLE_APP_UDHCPC
 static int dhcp_up(struct interface_defn_t *ifd, execfn *exec)
 {
+    int rv;
+
 #if ENABLE_FEATURE_IFUPDOWN_IP
 	/* ip doesn't up iface when it configures it (unlike ifconfig) */
 	if (!execute("ip link set[[ addr %hwaddress%]] %iface% up", ifd, exec))
@@ -542,9 +552,13 @@
 	if (!execute("ifconfig %iface%[[ hw %hwaddress%]] up", ifd, exec))
 		return 0;
 #endif
-	return execute("udhcpc -R -b -p /var/run/udhcpc.%iface%.pid "
-			"-i %iface%[[ -H %hostname%]][[ -c %clientid%]][[ -s %script%]][[ %udhcpc_opts%]]",
-			ifd, exec);
+    if (!(rv = execute("dhcp4.script %iface% [[-h %hostname%]] [[-l %clientid%]]", ifd, exec))){
+    	return execute("udhcpc -R -b -p /var/run/udhcpc.%iface%.pid "
+	    		"-i %iface%[[ -H %hostname%]][[ -c %clientid%]][[ -s %script%]][[ %udhcpc_opts%]]",
+    			ifd, exec); 
+    }else{
+        return rv;
+    }
 }
 #else
 static int dhcp_up(struct interface_defn_t *ifd UNUSED_PARAM,
