--- busybox-old/networking/udhcp/dhcpc.c	2008-11-10 01:27:58.000000000 +0800
+++ busybox-new/networking/udhcp/dhcpc.c	2010-05-25 03:30:58.250144290 +0800
@@ -126,7 +126,7 @@
 int udhcpc_main(int argc UNUSED_PARAM, char **argv)
 {
 	uint8_t *temp, *message;
-	char *str_c, *str_V, *str_h, *str_F, *str_r;
+	char *str_c, *str_V, *str_h, *str_F, *str_r, *str_I;
 	USE_FEATURE_UDHCP_PORT(char *str_P;)
 	llist_t *list_O = NULL;
 	int tryagain_timeout = 20;
@@ -154,6 +154,7 @@
 		"hostname\0"       Required_argument "H"
 		"fqdn\0"           Required_argument "F"
 		"interface\0"      Required_argument "i"
+        "interval\0"       Required_argument "I"
 		"now\0"            No_argument       "n"
 		"pidfile\0"        Required_argument "p"
 		"quit\0"           No_argument       "q"
@@ -195,8 +196,9 @@
 		OPT_O = 1 << 18,
 		OPT_o = 1 << 19,
 		OPT_f = 1 << 20,
+        OPT_I = 1 << 21,
 /* The rest has variable bit positions, need to be clever */
-		OPTBIT_f = 20,
+		OPTBIT_I = 21,
 		USE_FOR_MMU(              OPTBIT_b,)
 		USE_FEATURE_UDHCPC_ARPING(OPTBIT_a,)
 		USE_FEATURE_UDHCP_PORT(   OPTBIT_P,)
@@ -215,7 +217,7 @@
 	/* Cc: mutually exclusive; O: list; -T,-t,-A take numeric param */
 	opt_complementary = "c--C:C--c:O::T+:t+:A+";
 	USE_GETOPT_LONG(applet_long_options = udhcpc_longopts;)
-	opt = getopt32(argv, "c:CV:H:h:F:i:np:qRr:s:T:t:vSA:O:of"
+	opt = getopt32(argv, "c:CV:H:h:F:i:np:qRr:s:T:t:vSA:O:ofI:"
 		USE_FOR_MMU("b")
 		USE_FEATURE_UDHCPC_ARPING("a")
 		USE_FEATURE_UDHCP_PORT("P:")
@@ -225,6 +227,7 @@
 		, &discover_timeout, &discover_retries, &tryagain_timeout /* T,t,A */
 		, &list_O
 		USE_FEATURE_UDHCP_PORT(, &str_P)
+        , &str_I
 		);
 	if (opt & OPT_c)
 		client_config.clientid = alloc_dhcp_option(DHCP_CLIENT_ID, str_c, 0);
@@ -281,7 +284,12 @@
 		openlog(applet_name, LOG_PID, LOG_LOCAL0);
 		logmode |= LOGMODE_SYSLOG;
 	}
-
+    if (opt & OPT_I) {
+        client_config.interval = atoi(str_I);
+    }else{
+        printf("DHCP internal = 0 \n");
+        client_config.interval = 0;
+    }
 	/* Make sure fd 0,1,2 are open */
 	bb_sanitize_stdio();
 	/* Equivalent of doing a fflush after every \n */
@@ -360,7 +368,8 @@
 					send_discover(xid, requested_ip); /* broadcast */
 
 					timeout = discover_timeout;
-					packet_num++;
+                    if( discover_retries != 0xff) /* 0xFF is keeping retry */
+    					packet_num++;
 					continue;
 				}
  leasefail:
@@ -379,7 +388,7 @@
 					goto ret;
 				}
 				/* wait before trying again */
-				timeout = tryagain_timeout;
+    			timeout = tryagain_timeout + client_config.interval * 10;
 				packet_num = 0;
 				continue;
 			case RENEW_REQUESTED:
