--- uboot_old/include/env_default.h	2015-07-01 21:28:50.579257485 +0800
+++ uboot/include/env_default.h		2015-07-01 21:29:02.259257610 +0800
@@ -111,7 +111,7 @@
 	"hostname="	__stringify(CONFIG_HOSTNAME)	"\0"
 #endif
 #ifdef	CONFIG_BOOTFILE
-	"bootfile="	CONFIG_BOOTFILE			"\0"
+	"bootfile="	__stringify(CONFIG_BOOTFILE)			"\0"
 #endif
 #ifdef	CONFIG_LOADADDR
 	"loadaddr="	__stringify(CONFIG_LOADADDR)	"\0"
@@ -135,7 +135,18 @@
 #ifdef CONFIG_SPX_FEATURE_BMC_FIRMWARE_AUTO_RECOVERY
         "bootretry="  __stringify(CONFIG_SPX_FEATURE_BMC_FIRMWARE_REBOOT_RETRY_COUNT)    "\0"
 #endif
-
+#ifdef	CONFIG_FWU_ALL
+	"fwu_all="	__stringify(CONFIG_FWU_ALL)		"\0"
+#endif
+#ifdef	CONFIG_FWU_BOOT
+	"fwu_boot="	__stringify(CONFIG_FWU_BOOT)		"\0"
+#endif
+#ifdef	CONFIG_FWU_OS
+	"fwu_os="	__stringify(CONFIG_FWU_OS)		"\0"
+#endif
+#ifdef	CONFIG_BOARDID
+	"boardid="	__stringify(CONFIG_BOARDID)		"\0"
+#endif
 #if defined(CONFIG_PCI_BOOTDELAY) && (CONFIG_PCI_BOOTDELAY > 0)
 	"pcidelay="	__stringify(CONFIG_PCI_BOOTDELAY)"\0"
 #endif
